controllers/onload_controller.go

```diff
@@ -78,34 +90,515 @@ func (r *OnloadReconciler) Reconcile(
+	if err != nil {
+		log.Error(err, "Failed to add kmm label to nodes")
+		return ctrl.Result{}, err
+	} else if res != nil {
+		log.Info("Labelled nodes with kmm label. Re-queueing")
+		return *res, nil
+	}
 
-	ok, res, err := r.createAndAddModules(ctx, onload)
-	if !ok {
-		return res, err
+	res, err = r.createAndAddModules(ctx, onload)
+	if err != nil {
+		log.Info("Error creating module")
+		return ctrl.Result{}, err
+	} else if res != nil {
+		log.Info("Added onload module")
+		return *res, nil
 	}
 
-	log.Info("Adding Onload Device Plugin DaemonSet", "onload.Name",
-		onload.Name, "onload.Namespace", onload.Namespace)
+	res, err = r.addOnloadLabelsToNodes(ctx, onload)
+	if err != nil {
+		log.Error(err, "Failed to add onload label to nodes")
+		return ctrl.Result{}, err
+	} else if res != nil {
+		log.Info("Labelled nodes with onload label. Re-queueing")
+		return *res, nil
+	}
 
-	err = r.createDevicePluginDaemonSet(ctx, onload)
+	res, err = r.createDevicePluginDaemonSet(ctx, onload)
 	if err != nil {
 		log.Error(err, "Failed to create device plugin daemonset")
 		return ctrl.Result{}, err
+	} else if res != nil {
+		log.Info("Added Onload Device Plugin DaemonSet")
+		return *res, nil
+	}
+
+	res, err = r.handleUpdate(ctx, onload)
+	if err != nil {
+		log.Error(err, "Failed to handle updates")
+		return ctrl.Result{}, err
+	}
+	if res != nil {
+		log.Info("Onload CR is updating")
+		return *res, nil
 	}
 
 	return ctrl.Result{}, nil
 }
 
+const kmmModuleLabelPrefix = "kmm.node.kubernetes.io/version-module"
+const onloadModuleNameSuffix = "-onload-module"
+
+func kmmLabelName(name, namespace string) string {
+	return kmmModuleLabelPrefix + "." + namespace + "." + name + onloadModuleNameSuffix
+}
+
+const onloadLabelPrefix = "onload.amd.com/"
+
+func onloadLabelName(name, namespace string) string {
+	return onloadLabelPrefix + namespace + "." + name
+}
+
+func (r *OnloadReconciler) listNodesWithLabels(ctx context.Context, labelString string) (corev1.NodeList, error) {
+	selector, err := labels.Parse(labelString)
+	if err != nil {
+		return corev1.NodeList{}, err
+	}
+
+	nodes := corev1.NodeList{}
+	opt := client.MatchingLabelsSelector{Selector: selector}
+	err = r.List(ctx, &nodes, opt)
+	if err != nil {
+		return corev1.NodeList{}, err
+	}
+	return nodes, nil
+}
+
+func (r *OnloadReconciler) deleteLabels(ctx context.Context, namespacedName types.NamespacedName) error {
+	log := log.FromContext(ctx)
+
+	labelKey := kmmLabelName(namespacedName.Name, namespacedName.Namespace)
+
+	nodes, err := r.listNodesWithLabels(ctx, labelKey)
+	if err != nil {
+		return err
+	}
+
+	for _, node := range nodes.Items {
+		nodeCopy := node.DeepCopy()
+		delete(node.Labels, labelKey)
+		err := r.Patch(ctx, &node, client.MergeFrom(nodeCopy))
+		if err != nil {
+			log.Error(err, "Failed to patch node"+node.Name)
+			return err
+		}
+
+	}
+
+	labelKey = onloadLabelName(namespacedName.Name, namespacedName.Namespace)
```

Comment:
Shall we make `labelKey` a parameter and shorten the function as currently there are two identical blocks of code?